opnode/cmd/stateviz/main.go

```diff
@@ -0,0 +1,248 @@
+		return err
+	}
+	if err := json.Unmarshal(unquote(t.L2Head), &e.L2Head); err != nil {
+		return err
+	}
+	if err := json.Unmarshal(unquote(t.L2SafeHead), &e.L2SafeHead); err != nil {
+		return err
+	}
+	if err := json.Unmarshal(unquote(t.L2FinalizedHead), &e.L2FinalizedHead); err != nil {
+		return err
+	}
+	if err := json.Unmarshal(unquote(t.L1WindowBuf), &e.L1WindowBuf); err != nil {
+		return err
+	}
+	if e.L1WindowBuf == nil {
+		e.L1WindowBuf = make([]eth.BlockID, 0)
+	}
+	return nil
+}
+
+//go:embed assets
+var embeddedAssets embed.FS
+
+func main() {
+	flag.Parse()
+
+	log.Root().SetHandler(
+		log.LvlFilterHandler(log.LvlDebug, log.StreamHandler(os.Stdout, log.TerminalFormat(true))),
+	)
+
+	if *snapshot == "" {
+		log.Crit("missing required -snapshot flag")
+	}
+
+	sub, err := fs.Sub(embeddedAssets, "assets")
+	if err != nil {
+		log.Crit("Failed to open asset directory", "message", err)
+	}
+	assetFS = sub
+
+	go func() {
+		ticker := time.NewTicker(*refresh)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				// TODO: incremental load
+				log.Info("loading snapshot...")
+				if err := loadSnapshot(); err != nil {
+					log.Error("failed to load snapshot", "err", err)
+				}
+			}
+		}
+	}()
+
+	runServer()
+}
+
+func loadSnapshot() error {
+	file, err := os.Open(*snapshot)
+	if err != nil {
+		return fmt.Errorf("%w: failed to open snapshot file", err)
+	}
+	defer file.Close()
+
+	tempEntries := make(map[string][]SnapshotState)
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		var entry SnapshotState
+		if err := json.Unmarshal([]byte(scanner.Text()), &entry); err != nil {
+			return fmt.Errorf("%w: failed to decode snapshot log", err)
+		}
+
+		tempEntries[entry.EngineAddr] = append(tempEntries[entry.EngineAddr], entry)
+	}
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("%w: failed to scan snapshot file", err)
+	}
+
+	entriesMutex.Lock()
+	entries = tempEntries
+	entriesMutex.Unlock()
+
+	return nil
+}
+
+func runServer() {
+	l, err := net.Listen("tcp", *listenAddr)
+	if err != nil {
+		log.Crit("Failed to listen on address", "message", err)
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle("/", http.FileServer(http.FS(assetFS)))
+	mux.HandleFunc("/logs", makeGzipHandler(logsHandler))
+
+	log.Info("running webserver...")
+	if err := http.Serve(l, mux); err != nil && !errors.Is(err, http.ErrServerClosed) {
+		log.Crit("http server failed", "message", err)
+	}
+}
+
+type gzipResponseWriter struct {
+	io.Writer
+	http.ResponseWriter
+}
+
+func (w gzipResponseWriter) Write(b []byte) (int, error) {
+	return w.Writer.Write(b)
+}
+
+func makeGzipHandler(fn http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if !strings.Contains(r.Header.Get("Accept-Encoding"), "gzip") {
+			fn(w, r)
+			return
+		}
+		w.Header().Set("Content-Encoding", "gzip")
+		gz := gzip.NewWriter(w)
+		defer gz.Close()
```

Comment:
Does `close` get called too early here?

Reply:
I don't think so - it gets called when the handler function returns.